src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs

```diff
@@ -175,48 +178,96 @@ private void EnsureStartup()
 
         private RequestDelegate BuildApplication()
         {
-            if (ServerFactory == null)
+            try
             {
-                // Blow up if we don't have a server set at this point
-                if (ServerFactoryLocation == null)
+                EnsureApplicationServices();
+                EnsureServer();
+
+                var builderFactory = _applicationServices.GetRequiredService<IApplicationBuilderFactory>();
+                var builder = builderFactory.CreateBuilder(_serverInstance);
+                builder.ApplicationServices = _applicationServices;
+
+                var startupFilters = _applicationServices.GetService<IEnumerable<IStartupFilter>>();
+                var configure = Startup.ConfigureDelegate;
+                foreach (var filter in startupFilters)
                 {
-                    throw new InvalidOperationException("IHostingBuilder.UseServer() is required for " + nameof(Start) + "()");
+                    configure = filter.Configure(configure);
                 }
 
-                ServerFactory = _applicationServices.GetRequiredService<IServerLoader>().LoadServerFactory(ServerFactoryLocation);
-            }
-
-            _serverInstance = ServerFactory.Initialize(_config);
-            var builderFactory = _applicationServices.GetRequiredService<IApplicationBuilderFactory>();
-            var builder = builderFactory.CreateBuilder(_serverInstance);
-            builder.ApplicationServices = _applicationServices;
+                configure(builder);
 
-            var addresses = builder.ServerFeatures?.Get<IServerAddressesFeature>()?.Addresses;
-            if (addresses != null && !addresses.IsReadOnly)
+                return builder.Build();
+            }
+            catch (Exception ex)
             {
-                var port = _config[ServerPort];
-                if (!string.IsNullOrEmpty(port))
+                if (!_captureStartupErrors)
                 {
-                    addresses.Add("http://localhost:" + port);
+                    throw;
                 }
 
-                // Provide a default address if there aren't any configured.
-                if (addresses.Count == 0)
+                // EnsureApplicationServices may have failed due to a missing or throwing Startup class.
+                if (_applicationServices == null)
                 {
-                    addresses.Add("http://localhost:5000");
+                    _applicationServices = _applicationServiceCollection.BuildServiceProvider();
                 }
+
+                EnsureServer();
+
+                // Write errors to standard out so they can be retrieved when not in development mode.
+                Console.Out.WriteLine("Application startup exception: " + ex.ToString());
+                var logger = _applicationServices.GetRequiredService<ILogger<HostingEngine>>();
```

Comment:
Its like LINQ, nice to use in some cases but mostly easy to do something in efficient. It's worth trying to see why it's so slow. TBH tho, this code path is error path so it doesn't matter that much.